File: R/bert-nsp.R
create_nsp_examples <- function(documents, num_examples, seed = NULL) {
  if (!is.null(seed)) {
    set.seed(seed)
  }

  examples <- list()
  while (length(examples) < num_examples) {
    doc_idx <- sample(seq_along(documents), 1)
    document <- documents[[doc_idx]]

    if (length(document) < 2) {
      next
    }

    sent_idx <- sample(1:(length(document) - 1), 1)
    if (runif(1) < 0.5) {
      examples[[length(examples) + 1]] <- list(document[[sent_idx]], document[[sent_idx + 1]], 1)
    } else {
      if (length(documents) > 1) {
        random_doc_idx <- doc_idx
        while (random_doc_idx == doc_idx) {
          random_doc_idx <- sample(seq_along(documents), 1)
        }
        random_document <- documents[[random_doc_idx]]
      } else {
        random_document <- document
      }
      random_sent_idx <- sample(seq_along(random_document), 1)
      examples[[length(examples) + 1]] <- list(document[[sent_idx]], random_document[[random_sent_idx]], 0)
    }
  }

  examples[1:num_examples]
}

NSPHead <- setRefClass(
  "NSPHead",
  fields = list(
    W = "matrix",
    b = "numeric"
  ),
  methods = list(
    initialize = function(hidden_size) {
      W <<- matrix(rnorm(hidden_size * 2, sd = 0.02), nrow = hidden_size, ncol = 2)
      b <<- numeric(2)
    },
    forward = function(cls_hidden) {
      cls_hidden %*% W + b
    }
  )
)

softmax <- function(x) {
  exp_x <- exp(x - apply(x, 2, max))
  exp_x / rowSums(exp_x)
}

File: Julia/bert-nsp.jl
using Random

function create_nsp_examples(documents, num_examples::Int; seed=nothing)
    if seed !== nothing
        Random.seed!(seed)
    end

    examples = []
    while length(examples) < num_examples
        doc_idx = rand(1:length(documents))
        document = documents[doc_idx]

        if length(document) < 2
            continue
        end

        sent_idx = rand(1:(length(document) - 1))

        if rand() < 0.5
            push!(examples, (document[sent_idx], document[sent_idx + 1], 1))
        else
            if length(documents) > 1
                random_doc_idx = doc_idx
                while random_doc_idx == doc_idx
                    random_doc_idx = rand(1:length(documents))
                end
                random_document = documents[random_doc_idx]
            else
                random_document = document
            end
            random_sent_idx = rand(1:length(random_document))
            push!(examples, (document[sent_idx], random_document[random_sent_idx], 0))
        end
    end

    examples[1:num_examples]
end

mutable struct NSPHead
    W
    b
end

function NSPHead(hidden_size::Int)
    W = randn(hidden_size, 2) .* 0.02
    b = zeros(2)
    NSPHead(W, b)
end

function forward(head::NSPHead, cls_hidden)
    cls_hidden * head.W .+ head.b
end

softmax(x) = exp.(x .- maximum(x, dims=2)) ./ sum(exp.(x .- maximum(x, dims=2)), dims=2)
